File: go/election-day/election_day.go
package electionday

import (
	"fmt"
)

// NewVoteCounter returns a new vote counter with
// a given number of initial votes.
func NewVoteCounter(initialVotes int) *int {
	var counter *int

	counter = &initialVotes

	return counter
}

// VoteCount extracts the number of votes from a counter.
func VoteCount(counter *int) int {
	var votes int

	if counter != nil {
		votes = *counter
	} else {
		votes = 0
	}

	return votes
}

// IncrementVoteCount increments the value in a vote counter
func IncrementVoteCount(counter *int, increment int) {
	if counter != nil {
		*counter += increment
	} else {
		*counter = increment
	}
}

// NewElectionResult creates a new election result
func NewElectionResult(candidateName string, votes int) *ElectionResult {
	var result ElectionResult = ElectionResult{
		Name:  candidateName,
		Votes: votes,
	}

	return &result
}

// DisplayResult creates a message with the result to be displayed
func DisplayResult(result *ElectionResult) string {
	var message string

	message = fmt.Sprintf("%v (%v)", result.Name, result.Votes)

	return message
}

// DecrementVotesOfCandidate decrements by one the vote count of a candidate in a map
func DecrementVotesOfCandidate(results map[string]int, candidate string) {
	for key, value := range results {
		if key == candidate && value > 0 {
			results[key]--
		}
	}
}

File: cpp/election-day/election_day.cpp
#include <algorithm>
#include <string>
#include <vector>

namespace election {

using CadidateName = std::string;
using VoteCount = int;

// The election result struct is already created for you:
struct ElectionResult {
    // Name of the candidate
    CadidateName name{};

    // Number of votes the candidate has
    VoteCount votes{};
};

using ElectionSummary = std::vector<ElectionResult>;

// Task 1: vote_count()
// - takes a reference to an `ElectionResult` as an argument
// - will return the number of votes in the `ElectionResult`
VoteCount vote_count(ElectionResult &result) { return result.votes; }

// Task 2: increment_vote_count()
// - takes a reference to an `ElectionResult` as an argument
// - and a number of votes (VoteCount)
// - increments the `ElectionResult` by that number of votes
void increment_vote_count(ElectionResult &result, VoteCount count) {
    result.votes += count;
}

// Task 3: determine_result()
// - receives the reference to a final_count
// - returns a reference to the `ElectionResult` of the new president
// - it also changes the name of the winner by prefixing it with "President"
// - the final count is given in the form of a `reference` to
// `ElectionSummary`, a vector with `ElectionResults` of all the
// participating candidates
ElectionResult &determine_result(ElectionSummary &final_count) {
    int index{0};
    int counter{0};
    VoteCount highest_score{0};

    // index is a reference so we can access the "loop" value after
    // std::transform runs
    auto find_winner_op = [counter, &index, highest_score](
                              ElectionResult result) mutable -> ElectionResult {
        if (result.votes > highest_score) {
            highest_score = result.votes;
            index = counter;
        }

        counter += 1;

        return result;
    };

    std::transform(final_count.begin(), final_count.end(), final_count.begin(),
                   find_winner_op);

    final_count.at(index).name = "President " + final_count.at(index).name;

    return final_count.at(index);
}

} // namespace election
